### ISSUE TITLE
STRtree with empty geometries segfaults

### ISSUE THREAD

**Issue by zerebubuth** [Toblerity/Shapely#345]
When constructing an STRtree with an empty geometry, any query against that will cause a segfault as GEOS attempts to dereference null bounds for the empty geometry. It seems like empty geometries should be skipped?

Example:

```python
>>> from shapely.geometry import Polygon
>>> from shapely.strtree import STRtree
>>> polys = [ Polygon() ]
>>> tree = STRtree(polys)
>>> tree.query(Polygon([(0,0),(1,1),(2,0),(0,0)]))
Segmentation fault (core dumped)
```

(Tested with Python 2.7.6 on Ubuntu 14.04.3 with Shapely 1.5.13)

**Comment by snorfalorpagus** [comment]
Looks like this is an issue for empty geometries only. Invalid geometries don't cause a segfault. It should be possible to check for this before they are inserted and raise an exception if required.

**Comment by snorfalorpagus** [comment]
It seems to be legitimate to create an STRtree with no geometries, i.e. `tree = STRtree([])`. What should the behaviour of `tree = STRtree([Polygon()])` be? Raise an exception, or return an empty tree? Also, does it matter if the GEOS STRtree is initalised with the wrong size. This will happen if we use `len(geoms)` but then don't insert all of the geometries.

**Comment by sgillies** [comment]
@snorfalorpagus let's do this:

- continue to allow create of empty trees since that isn't a reported problem
- add only non-empty geoms to the tree and make sure the tree doesn't have undefined geoms

By that i mean filter the `STRtree()` input sequence and make a new list of non-empty geoms. Use the length of that list when initializing. There's some overhead to making that new list but it's not much, it doesn't allocate new memory for geometry objects.

**Comment by snorfalorpagus** [comment]
Closed by #348.
